### DISCUSSION TITLE
PNPM Support for Expo React-Native

### DISCUSSION THREAD

**Discussion by ImBIOS** [expo/expo#24031] [Proposals and Ideas]
**Objective:**

The aim of this discussion is to consider and evaluate the possibility of integrating `pnpm` support for the Expo React-Native platform.

---

**Background:**

With the growing demand for efficient package management solutions, `pnpm` has emerged as a robust and fast alternative to `npm` and `yarn`. It offers benefits like:

- **Efficient storage:** By using a content-addressable filesystem to store all files from all module directories in a single place.
- **Fast:** Due to its efficient approach in handling node_modules, installations are usually faster.
- **Strict linking:** It creates a hard link between the project's node_modules folder and the location where the package is stored. This means, if you have a single version of a package used in multiple projects, it'll only be stored once on the disk.

Given these advantages, it is evident that having `pnpm` support can enhance the developer experience on Expo React-Native.

---

**Proposal:**

1. **Initial Investigation:** Analyze the Expo React-Native codebase to determine any potential hurdles or dependencies that would need to be addressed for `pnpm` support.
  
2. **Community Feedback:** Solicit feedback from the Expo community to understand how many developers are using or are interested in using `pnpm`. This will help gauge the demand and prioritize the integration.
  
3. **Documentation:** Once implemented, provide thorough documentation and guidelines on how developers can switch to or utilize `pnpm` with Expo React-Native.

4. **Testing:** Before officially launching the support, create a beta version for the community. This will help in understanding any issues or challenges that might arise in different development environments.

---

**Benefits:**

- **Performance:** Faster and more efficient package installations.
- **Disk Space:** Reduced disk space usage due to efficient storage.
- **Broader Audience:** By supporting more package managers, we can cater to a larger developer audience and be more inclusive.

---

**Conclusion:**

The integration of `pnpm` can be a strategic move for Expo React-Native, given the array of benefits it brings. It would be enlightening to hear thoughts, feedback, and potential concerns from the community regarding this proposal.

---

*Please share your opinions, insights, and any potential challenges you foresee. Your input is invaluable in driving Expo React-Native towards an enhanced developer experience.*

**Comment by jeetgr** [comment]
I only have a daily mobile data limit of 2GB on my SIM card because there's no broadband connection available in my area. While working with React and Next.js, I use `pnpm` as a package manager, which helps save some data for me. However, in React-Native, I can no longer use `pnpm`, and this is not ideal for me. The 2GB daily limit is insufficient, and the charges from the SIM provider are high, making it difficult for me to afford an expensive data plan.

**Comment by anshuman** [comment]
Yup it's a great idea!
cc: @expo

**Comment by LukaRusadze** [comment]
Looking at the current state of react native, pnpm seems to be working without `node-linker=hoisted`. Expo's tooling seems more incompatible than react native itself. Looking at `@expo/cli` there seems to be some phantom dependencies (for example: `metro/src/lib/TerminalReporter` [#30143](https://github.com/expo/expo/issues/30143), happens because metro isnt a dependency of @expo/cli), `expo-router` also has an issue with `@babel/runtime/helpers/interopRequireDefault`

Maybe @EvanBacon can help out with this? Only reason I'm asking you is because you seem to be the one who worked with both of these projects/packages
